File: lexer/lexer.go
package lexer

import (
	"bytes"
)

type Lexer struct {
	inputChan    chan *Input
	currentInput *Input
	Line         int
	Column       int
	position     int // current position in input (points to current char)
	nextPosition int
	currentChar  byte // current char under examination
	outputChan   chan Token
}

func New(input *Input) *Lexer {
	l := &Lexer{inputChan: make(chan *Input, 100), outputChan: make(chan Token, 100), Line: 1, Column: -1}

	l.Feed(input)

	l.readChar()

	return l
}

func (l *Lexer) Snapshot(token Token) string {
	start := token.Pos - 30
	end := token.Pos + 5

	if start < 0 {
		start = 0
	}

	if end > len(l.currentInput.content) {
		end = len(l.currentInput.content)
	}

	var buf bytes.Buffer

	buf.WriteString(l.currentInput.content[start:end])
	buf.WriteString("\n")

	for i := 0; i < token.Pos-start; i++ {
		buf.WriteString(" ")
	}
	buf.WriteString("^")

	return buf.String()
}

func (l *Lexer) Feed(input *Input) {
	if l.currentInput == nil {
		l.currentInput = input
		return
	}

	l.inputChan <- input
}

func (l *Lexer) Run() {
	go func() {
		var tok Token

		for tok.Type != EOF {

			l.eatWhitespace()

			switch l.currentChar {
			case '#':
				for l.currentChar != '\n' {
					l.readChar()
				}
				continue
			case '=':
				if l.peekChar() == '=' {
					char := l.currentChar
					l.readChar()
					tok = Token{Type: EQ, Literal: string(char) + string(l.currentChar)}
				} else {
					tok = l.newToken(ASSIGN, l.currentChar)
				}
			case '!':
				if l.peekChar() == '=' {
					char := l.currentChar
					l.readChar()
					tok = Token{Type: NOT_EQ, Literal: string(char) + string(l.currentChar)}
				} else {
					tok = l.newToken(BANG, l.currentChar)
				}
			case '<':
				if l.peekChar() == '=' {
					char := l.currentChar
					l.readChar()
					tok = Token{Type: LT_OR_EQ, Literal: string(char) + string(l.currentChar)}
				} else {
					if l.peekChar() == '<' {
						char := l.currentChar
						l.readChar()
						tok = Token{Type: APPEND, Literal: string(char) + string(l.currentChar)}
					} else {
						tok = l.newToken(LT, l.currentChar)
					}
				}
			case '>':
				if l.peekChar() == '=' {
					char := l.currentChar
					l.readChar()
					tok = Token{Type: GT_OR_EQ, Literal: string(char) + string(l.currentChar)}
				} else {
					tok = l.newToken(GT, l.currentChar)
				}
			case ';':
				tok = l.newToken(SEMICOLON, l.currentChar)
			case '(':
				tok = l.newToken(LPAREN, l.currentChar)
			case ')':
				tok = l.newToken(RPAREN, l.currentChar)
			case ',':
				tok = l.newToken(COMMA, l.currentChar)
			case '+':
				tok = l.newToken(PLUS, l.currentChar)
			case '-':
				tok = l.newToken(MINUS, l.currentChar)
			case '/':
				tok = l.newToken(SLASH, l.currentChar)
			case '*':
				tok = l.newToken(ASTERISK, l.currentChar)
			case '{':
				tok = l.newToken(LBRACE, l.currentChar)
			case '}':
				tok = l.newToken(RBRACE, l.currentChar)
			case '[':
				tok = l.newToken(LBRACKET, l.currentChar)
			case ']':
				tok = l.newToken(RBRACKET, l.currentChar)
			case ':':
				tok = l.newToken(COLON, l.currentChar)
			case '.':
				tok = l.newToken(DOT, l.currentChar)
			case '"':
				tok.Type = STRING
				tok.Literal = l.readString()
			case '&':
				if l.peekChar() == '&' {
					char := l.currentChar
					l.readChar()

					if l.peekChar() == '=' {
						secondChar := l.currentChar
						l.readChar()
						tok = Token{Type: BOOL_AND_ASSIGN, Literal: string(char) + string(secondChar) + string(l.currentChar)}
					} else {
						tok = Token{Type: BOOL_AND, Literal: string(char) + string(l.currentChar)}
					}
				} else {
					tok = l.newToken(BIT_AND, l.currentChar)
				}
			case '|':
				if l.peekChar() == '|' {
					char := l.currentChar
					l.readChar()

					if l.peekChar() == '=' {
						secondChar := l.currentChar
						l.readChar()
						tok = Token{Type: BOOL_OR_ASSIGN, Literal: string(char) + string(secondChar) + string(l.currentChar)}
					} else {
						tok = Token{Type: BOOL_OR, Literal: string(char) + string(l.currentChar)}
					}
				} else {
					tok = l.newToken(BIT_OR, l.currentChar)
				}
			case '@':
				tok.Type = INSTANCE_VAR

				char := l.currentChar

				l.readChar()

				tok.Literal = string(char) + l.readIdentifier()

				l.sendToken(tok)
				continue
			case 0:
				tok.Literal = ""
				tok.Type = EOF
			default:
				if isLetter(l.currentChar) {
					tok.Pos = l.position
					tok.Column = l.Column
					tok.Line = l.Line
					tok.Literal = l.readIdentifier()
					tok.Type = lookupIdent(tok.Literal)
					l.sendToken(tok)
					continue
				} else if isDigit(l.currentChar) {
					tok.Pos = l.position
					tok.Column = l.Column
					tok.Line = l.Line
					tok.Type = INT
					tok.Literal = l.readNumber()
					l.sendToken(tok)
					continue
				} else {
					tok = l.newToken(ILLEGAL, l.currentChar)
				}
			}

			l.readChar()

			l.sendToken(tok)
		}

		l.sendToken(tok)
	}()
}

func (l *Lexer) Close() {
	close(l.inputChan)
	close(l.outputChan)
}

func (l *Lexer) sendToken(token Token) {
	l.outputChan <- token
}

func (l *Lexer) NextToken() Token {
	return <-l.outputChan
}

func (l *Lexer) newToken(tokenType TokenType, ch byte) Token {
	return Token{Type: tokenType, Literal: string(ch), Line: l.Line, Column: l.Column, Pos: l.position}
}

func (l *Lexer) readChar() {
	l.currentChar = l.peekChar()

	l.position = l.nextPosition
	l.nextPosition += 1

	l.Column += 1
}

func (l *Lexer) peekChar() byte {
	if l.currentInput == nil || l.nextPosition >= len(l.currentInput.content) {
		select {
		case nextInput := <-l.inputChan:
			l.currentInput = nextInput
		default:
			return 0
		}
		return 0
	} else {
		return l.currentInput.content[l.nextPosition]
	}
}

func (l *Lexer) eatWhitespace() {
	for l.currentChar == ' ' || l.currentChar == '\t' || l.currentChar == '\n' || l.currentChar == '\r' {
		isNewLine := l.currentChar == '\n'

		l.readChar()

		if isNewLine {
			l.Column = 1
			l.Line += 1
		}
	}
}

func (l *Lexer) readNumber() string {
	position := l.position
	for isDigit(l.currentChar) || l.currentChar == '_' {
		l.readChar()
	}
	return l.currentInput.content[position:l.position]
}

func (l *Lexer) readString() string {
	position := l.position + 1
	for {
		l.readChar()
		if l.currentChar == '"' || l.currentChar == 0 {
			break
		}
	}
	return l.currentInput.content[position:l.position]
}

func isDigit(ch byte) bool {
	return '0' <= ch && ch <= '9'
}

func (l *Lexer) readIdentifier() string {
	position := l.position
	for isLetter(l.currentChar) {
		l.readChar()
	}
	return l.currentInput.content[position:l.position]
}

func isLetter(ch byte) bool {
	return 'a' <= ch && ch <= 'z' || 'A' <= ch && ch <= 'Z' || ch == '_'
}

File: src/lexer.rs
use std::str;

mod char_classifiers;

pub mod input;
pub mod token;

pub struct Lexer {
    input: input::Input,
    line: i64,
    column: i64,
    position: i64, // current position in input (points to current char)
    next_position: i64,
    current_char: char, // current char under examination
}

impl Lexer {
    pub fn new(input: input::Input) -> Lexer {
        let mut lexer = Lexer {
            input,
            line: 1,
            column: 0,
            position: 0,
            next_position: 0,
            current_char: char::from(0),
        };

        lexer.read_char();

        return lexer;
    }

    pub fn next_token(&mut self) -> token::Token {
        self.eat_whitespace();

        let mut tok: token::Token = self.new_illegal_token();

        match self.current_char {
            '\u{0}' => tok = token::Token::Eof(self.new_token_data_with_current_char()),
            '#' => {
                while self.current_char != '\n' {
                    self.read_char();
                }
                tok = self.next_token()
            }
            '\n' => {
                tok = token::Token::Newline(self.new_token_data_with_current_char());
                self.line += 1;
                self.column = 0;
            }
            ';' => tok = token::Token::Semicolon(self.new_token_data_with_current_char()),
            '(' => tok = token::Token::LeftParen(self.new_token_data_with_current_char()),
            ')' => tok = token::Token::RightParen(self.new_token_data_with_current_char()),
            '[' => tok = token::Token::LeftBracket(self.new_token_data_with_current_char()),
            ']' => tok = token::Token::RightBracket(self.new_token_data_with_current_char()),
            '{' => tok = token::Token::LeftBrace(self.new_token_data_with_current_char()),
            '}' => tok = token::Token::RightBrace(self.new_token_data_with_current_char()),
            ',' => tok = token::Token::Comma(self.new_token_data_with_current_char()),
            '.' => tok = token::Token::Dot(self.new_token_data_with_current_char()),
            '=' => {
                if self.peek_char() == '=' {
                    tok = token::Token::Equals(self.new_token_data_with_current_and_peek_char());
                } else {
                    tok = token::Token::Assign(self.new_token_data_with_current_char());
                }
            }
            '!' => {
                if self.peek_char() == '=' {
                    tok = token::Token::NotEquals(self.new_token_data_with_current_and_peek_char());
                } else {
                    tok = token::Token::Bang(self.new_token_data_with_current_char());
                }
            }
            '&' => {
                if self.peek_char() == '&' {
                    let mut tok_data = self.new_token_data_with_current_and_peek_char();

                    if self.peek_char() == '=' {
                        self.read_char();
                        tok_data.literal.push(self.current_char);

                        tok = token::Token::BooleanAndAssign(tok_data);
                    } else {
                        tok = token::Token::BooleanAnd(tok_data);
                    }
                } else {
                    tok = token::Token::BitAnd(self.new_token_data_with_current_char());
                }
            }
            '|' => {
                if self.peek_char() == '|' {
                    let mut tok_data = self.new_token_data_with_current_and_peek_char();

                    if self.peek_char() == '=' {
                        self.read_char();
                        tok_data.literal.push(self.current_char);

                        tok = token::Token::BooleanOrAssign(tok_data);
                    } else {
                        tok = token::Token::BooleanOr(tok_data);
                    }
                } else {
                    tok = token::Token::BitOr(self.new_token_data_with_current_char());
                }
            }
            '+' => tok = token::Token::Plus(self.new_token_data_with_current_char()),
            '-' => tok = token::Token::Minus(self.new_token_data_with_current_char()),
            '/' => tok = token::Token::Slash(self.new_token_data_with_current_char()),
            '*' => tok = token::Token::Asterisk(self.new_token_data_with_current_char()),
            '<' => {
                if self.peek_char() == '<' {
                    tok = token::Token::Append(self.new_token_data_with_current_and_peek_char());
                } else if self.peek_char() == '=' {
                    tok = token::Token::LessThanOrEquals(
                        self.new_token_data_with_current_and_peek_char(),
                    );
                } else {
                    tok = token::Token::LessThan(self.new_token_data_with_current_char());
                }
            }
            '>' => {
                if self.peek_char() == '=' {
                    tok = token::Token::GreaterThanOrEquals(
                        self.new_token_data_with_current_and_peek_char(),
                    );
                } else {
                    tok = token::Token::GreaterThan(self.new_token_data_with_current_char());
                }
            }
            '$' => {
                return self.lex_global_ident();
            }
            '@' => {
                return self.lex_instance_var();
            }
            '"' => {
                return self.lex_string();
            }
            _ => {
                if char_classifiers::is_letter(self.current_char) {
                    return self.lex_identifier();
                } else if char_classifiers::is_digit(self.current_char) {
                    return self.lex_numeric();
                }
            }
        }

        self.read_char();

        tok
    }

    fn lex_identifier(&mut self) -> token::Token {
        let mut tok_data = self.new_token_data("".to_string());

        tok_data.literal = self.read_identifier();

        match tok_data.literal.as_str() {
            "class" => token::Token::Class(tok_data),
            "module" => token::Token::Module(tok_data),
            "def" => token::Token::Def(tok_data),
            "do" => token::Token::Do(tok_data),
            "begin" => token::Token::Begin(tok_data),
            "rescue" => token::Token::Rescue(tok_data),
            "ensure" => token::Token::Ensure(tok_data),
            "if" => token::Token::If(tok_data),
            "else" => token::Token::Else(tok_data),
            "end" => token::Token::End(tok_data),
            "self" => token::Token::Selff(tok_data),
            "return" => token::Token::Return(tok_data),
            "true" => token::Token::True(tok_data),
            "false" => token::Token::False(tok_data),
            "nil" => token::Token::Nil(tok_data),
            _ => token::Token::Ident(tok_data),
        }
    }

    fn lex_numeric(&mut self) -> token::Token {
        let mut tok_data = self.new_token_data("".to_string());

        tok_data.literal = self.read_number();

        if self.current_char == '.' {
            self.read_char();

            let decimals = self.read_number();

            tok_data.literal.push('.');
            tok_data.literal.push_str(&*decimals);

            token::Token::Float(tok_data)
        } else {
            token::Token::Int(tok_data)
        }
    }

    fn lex_instance_var(&mut self) -> token::Token {
        let mut tok_data = self.new_token_data_with_current_char();

        self.read_char();

        tok_data.literal.push_str(&*self.read_identifier());

        token::Token::InstanceVar(tok_data)
    }

    fn lex_global_ident(&mut self) -> token::Token {
        let mut tok_data = self.new_token_data_with_current_char();

        self.read_char();

        if self.current_char == ':' {
            tok_data.literal.push(self.current_char);
            self.read_char()
        } else {
            tok_data.literal.push_str(&*self.read_identifier());
        }

        token::Token::GlobalIdent(tok_data)
    }

    fn lex_string(&mut self) -> token::Token {
        let mut tok_data = self.new_token_data("".to_string());

        let position = self.next_position;

        self.read_char();

        while self.current_char != '"' {
            self.read_char();
        }

        let string_bytes =
            &self.input.content.as_bytes()[position as usize..self.position as usize];

        tok_data
            .literal
            .push_str(str::from_utf8(string_bytes).unwrap());

        self.read_char();

        token::Token::String(tok_data)
    }

    fn read_char(&mut self) {
        self.current_char = self.peek_char();

        self.position = self.next_position;
        self.next_position += 1;
        self.column += 1;
    }

    fn peek_char(&self) -> char {
        if self.next_position >= self.input.content.len().try_into().unwrap() {
            char::from(0)
        } else {
            self.input
                .content
                .chars()
                .nth(self.next_position as usize)
                .unwrap()
        }
    }

    fn eat_whitespace(&mut self) {
        while self.current_char == ' ' || self.current_char == '\t' || self.current_char == '\r' {
            self.read_char()
        }
    }

    fn read_identifier(&mut self) -> String {
        let position = self.position;

        while char_classifiers::is_letter(self.current_char) {
            self.read_char()
        }

        if self.current_char == '?' || self.current_char == '!' {
            self.read_char()
        }

        let ident_bytes = &self.input.content.as_bytes()[position as usize..self.position as usize];

        String::from(str::from_utf8(ident_bytes).unwrap())
    }

    fn read_number(&mut self) -> String {
        let position = self.position;

        while char_classifiers::is_digit(self.current_char) || self.current_char == '_' {
            self.read_char();
        }

        let ident_bytes = &self.input.content.as_bytes()[position as usize..self.position as usize];

        String::from(str::from_utf8(ident_bytes).unwrap())
    }

    fn new_illegal_token(&self) -> token::Token {
        token::Token::Illegal(self.new_token_data(self.current_char.to_string()))
    }

    fn new_token_data(&self, literal: String) -> token::TokenData {
        token::TokenData {
            literal,
            line: self.line,
            column: self.column,
            pos: self.position,
        }
    }

    fn new_token_data_with_current_char(&self) -> token::TokenData {
        self.new_token_data(self.current_char.to_string())
    }

    fn new_token_data_with_current_and_peek_char(&mut self) -> token::TokenData {
        let mut tok_data = self.new_token_data_with_current_char();

        self.read_char();

        tok_data.literal.push(self.current_char);

        tok_data
    }
}
